fig/cli/config.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import unicode_literals
+from __future__ import absolute_import
+
+import six
+import errno
+import yaml
+import os
+import re
+
+from . import errors
+
+
+class Config:
+    """
+    Wraps our YAML configuration to introduce support for environment variables and defaults
+    """
+    def __init__(self, config_dict):
+        """
+        Takes a dict (usually the output of a yaml file) and stores it.
+        Importantly, it recurses any structured values to turn them into configs as well.
+        We run string interpolations on stringy values if they match the ${ENV_VAR} pattern.
+
+        :param config_dict: a dict with structured configuration details
+        :type config_dict: dict
+        """
+
+        self.config_dict = dict([(key, Config(value)) if type(value) == dict else (key, value)
+                                 for key, value in config_dict.items()])
+
+    def as_dict(self):
+        """
+        Recursively forces a preprocessed config back into dict type
+        This is much easier than rewriting all client code to use a different class,
+        while providing us with a optional upgrade path (implementing Config directly where possible)
+
+        :return: the config as represented in a dict format
+        :rtype: dict
+        """
+
+        new_dict = {}
+        for key, value in self.config_dict.items():
+            if value.__class__ == Config:
+                new_dict[key] = value.as_dict()
+            elif type(value) == list:
+                new_dict[key] = [self.fully_interpolated(x) for x in value]
+            elif type(value) == dict:
+                new_dict[key] = dict([(subkey, self.fully_interpolated(subvalue))
+                                      for subkey, subvalue in value.items()])
+            else:
+                new_dict[key] = self.fully_interpolated(value)
+        return new_dict
+
+    @classmethod
+    def fully_interpolated(cls, value):
+        """
+        Matches our environment variable pattern, replaces the value with the value in the env
+
+        A future improvement here would be to also reference a separate dictionary that keeps track of
+        configurable variables via flat file.
+
+        :param value: any value that is reachable by a key in a dict represented by a yaml file
+        :return: the value itself if not a string with an environment variable, otherwise the value specified in the env
+        """
+        if not (isinstance(value, str) or type(value) == unicode):
+            return value
+
+        env_regex = r'\$\{(?P<env_var>[^\}^:]+)(:(?P<default_val>[^\}]+))?\}'
```

Comment:
It appears that using string.Template would also require regular expressions, so it wouldn't be much better than what's proposed. And string.Formatter appears to bury most of its logic in str and unicode built-ins.

I know that regular expressions can be tricky, but they are more performant and -- when used sparingly -- can be easier to maintain than creating a full-scale parser for this use case. 

Let me know if there's something I can do to help make you feel a bit better about this implementation detail.